src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py

```diff
@@ -0,0 +1,1781 @@
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.bias is not None:
+                module.bias.data.zero_()
+
+    def apply_weight_norm(self):
+        nn.utils.weight_norm(self.conv_pre)
+        for layer in self.upsampler:
+            nn.utils.weight_norm(layer)
+        for layer in self.resblocks:
+            layer.apply_weight_norm()
+        nn.utils.weight_norm(self.conv_post)
+
+    def remove_weight_norm(self):
+        nn.utils.remove_weight_norm(self.conv_pre)
+        for layer in self.upsampler:
+            nn.utils.remove_weight_norm(layer)
+        for layer in self.resblocks:
+            layer.remove_weight_norm()
+        nn.utils.remove_weight_norm(self.conv_post)
+
+    def forward(self, spectrogram: torch.FloatTensor) -> torch.FloatTensor:
+        r"""
+        Converts a log-mel spectrogram into a speech waveform. Passing a batch of log-mel spectrograms returns a batch
+        of speech waveforms. Passing a single, un-batched log-mel spectrogram returns a single, un-batched speech
+        waveform.
+
+        Args:
+            spectrogram (`torch.FloatTensor`):
+                Tensor containing the log-mel spectrograms. Can be batched and of shape `(batch_size, sequence_length,
+                config.model_in_dim)`, or un-batched and of shape `(sequence_length, config.model_in_dim)`.
+
+        Returns:
+            `torch.FloatTensor`: Tensor containing the speech waveform. If the input spectrogram is batched, will be of
+            shape `(batch_size, num_frames,)`. If un-batched, will be of shape `(num_frames,)`.
+        """
+        if self.config.normalize_before:
+            spectrogram = (spectrogram - self.mean) / self.scale
+
+        is_batched = spectrogram.dim() == 3
+        if not is_batched:
+            spectrogram = spectrogram.unsqueeze(0)
+
+        hidden_states = spectrogram.transpose(2, 1)
+
+        hidden_states = self.conv_pre(hidden_states)
+        for i in range(self.num_upsamples):
+            hidden_states = nn.functional.leaky_relu(hidden_states, self.config.leaky_relu_slope)
+            hidden_states = self.upsampler[i](hidden_states)
+
+            res_state = self.resblocks[i * self.num_kernels](hidden_states)
+            for j in range(1, self.num_kernels):
+                res_state += self.resblocks[i * self.num_kernels + j](hidden_states)
+            hidden_states = res_state / self.num_kernels
+
+        hidden_states = nn.functional.leaky_relu(hidden_states)
+        hidden_states = self.conv_post(hidden_states)
+        hidden_states = torch.tanh(hidden_states)
+
+        if not is_batched:
+            # remove batch dim and collapse tensor to 1-d audio waveform
+            waveform = hidden_states.squeeze(0).transpose(1, 0).view(-1)
+        else:
+            # remove seq-len dim since this collapses to 1
+            waveform = hidden_states.squeeze(1)
+
+        return waveform
```

Comment:
Would be nice to have a model called `FastSpeech2ConformerWithHiFiGan` that uses the FS2 model as a base model and the HiFi Gan as a head on top -> this way a user only has to do `from_pretrained` once to get the weights of both models and the vocoder is used directly in the forward pass (better API IMO)

Reply:
To my knowledge this would be the first model in the library to have a head that is itself a self-contained model in the library. 

Given this, I’m not certain what’s best to do for the config for this combined model. If we have an entirely separate config that combines attributes of both the model and vocoder, we’ll have to update two separate configs when changes are made.

The only reason the vocoder is separate in the first place is to follow the `SpeechT5` pattern and since different vocoders could theoretically be used, but maybe we should just move it into the base model and determine whether to pass the outputs through the vocoder using a config attribute or forward pass arg?

Reply:
Hi @connor-henderson, 
`SpeechT5` pattern is not the typical transformers pattern with the `generate_speech` method, and I think we're moving towards a more consistent library so I guess we should not strictly copy `SpeechT5` pattern.

There are at least two models that I know of, [MusicGen](https://github.com/huggingface/transformers/pull/24109) and [Bark](https://github.com/huggingface/transformers/pull/24086), that uses self-containted models as head.

You can get inspiration from both models for properly initiating the configurations and the models!

Reply:
Ah thanks @ylacombe, hadn't seen those new models yet the `BarkConfig` with `from_sub_model_configs` is exactly what I was looking for